src/Controller/Component/SecurityComponent.php

```
@@ -281,18 +318,83 @@ protected function _validatePost(Controller $controller)
         if (empty($controller->request->data)) {
             return true;
         }
+        $token = $this->_validToken($controller);
+        $hashParts = $this->_hashParts($controller);
+        $check = Security::hash(implode('', $hashParts), 'sha1');
+
+        if ($token === $check) {
+            return true;
+        }
+
+        if (Configure::read('debug')) {
+            $msg = $this->_debugPostTokenNotMatching($controller, $hashParts);
+            throw new SecurityException($msg);
+        }
+
+        return false;
+    }
+
+    /**
+     * Check if token is valid
+     *
+     * @param \Cake\Controller\Controller $controller Instantiating controller
+     * @throws \Cake\Controller\Exception\SecurityException
+     * @return string fields token
+     */
+    protected function _validToken(Controller $controller)
+    {
         $check = $controller->request->data;
 
-        if (!isset($check['_Token']) ||
-            !isset($check['_Token']['fields']) ||
-            !isset($check['_Token']['unlocked'])
-        ) {
-            return false;
+        $message = '\'%s\' was not found in request data.';
+        if (!isset($check['_Token'])) {
+            throw new AuthSecurityException(sprintf($message, '_Token'));
+        }
+        if (!isset($check['_Token']['fields'])) {
+            throw new AuthSecurityException(sprintf($message, '_Token.fields'));
+        }
+        if (!isset($check['_Token']['unlocked'])) {
+            throw new AuthSecurityException(sprintf($message, '_Token.unlocked'));
         }
+        if (Configure::read('debug') && !isset($check['_Token']['debug'])) {
+            throw new SecurityException(sprintf($message, '_Token.debug'));
+        }
```

Should you also throw an exception if someone includes debug data in prod mode?